File: Laba_2Python/MatrixData.py
import numpy as np

class MyMatrix:
    def __init__(self, source):
        self.matrix = None
        self._cached_determinant = None

        if isinstance(source, MyMatrix):
            self.matrix = np.array(source.matrix, copy=True)

        elif isinstance(source, np.ndarray):
            self.matrix = source.copy()

        elif isinstance(source, list):
            if all(isinstance(row, list) for row in source):
                self._validate_jagged_array(source)
                self.matrix = np.array(source, dtype=float)

            elif all(isinstance(row, str) for row in source):
                self.matrix = self._from_string_list(source)

            else:
                raise ValueError("Invalid source data for matrix")

        elif isinstance(source, str):
            self.matrix = self._from_string(source)

        else:
            raise ValueError("Unsupported type for matrix initialization")

    @property
    def height(self):
        return self.matrix.shape[0]

    @property
    def width(self):
        return self.matrix.shape[1]
    
    def get_height(self):
        return self.matrix.shape[0]
    
    def get_width(self):
        return self.matrix.shape[1]

    def __getitem__(self, indices):
        i, j = indices
        self._validate_indexes(i, j)
        return self.matrix[i, j]

    def __setitem__(self, indices, value):
        i, j = indices
        self._validate_indexes(i, j)
        self._cached_determinant = None
        self.matrix[i, j] = value

    def get_element(self, i, j):
        self._validate_indexes(i, j)
        return self.matrix[i, j]

    def set_element(self, i, j, value):
        self._validate_indexes(i, j)
        self._cached_determinant = None
        self.matrix[i, j] = value

    def _validate_indexes(self, i, j):
        if not (0 <= i < self.height):
            raise IndexError(f"Row index {i} is out of range [0, {self.height})")
        if not (0 <= j < self.width):
            raise IndexError(f"Column index {j} is out of range [0, {self.width})")

    def _validate_jagged_array(self, jagged_array):
        if not jagged_array or any(len(row) != len(jagged_array[0]) for row in jagged_array):
            raise ValueError("Matrix must be rectangular")

    def _from_string_list(self, rows):
        if not rows:
            raise ValueError("Array cannot be empty")
        matrix = []
        for i, row in enumerate(rows):
            try:
                numbers = list(map(float, row.split()))
                if len(matrix) > 0 and len(numbers) != len(matrix[0]):
                    raise ValueError("All rows must have the same number of elements")
                matrix.append(numbers)
            except ValueError:
                raise ValueError(f"Invalid number format at row {i}")
        return np.array(matrix)

    def _from_string(self, matrix_string):
        rows = matrix_string.strip().splitlines()
        return self._from_string_list(rows)

    def __str__(self):
        return "\n".join("\t".join(map(str, row)) for row in self.matrix)

File: Laba_2/Laba_2/MatrixData.cs
using System.Globalization;


namespace Laba_2
{
    public partial class MyMatrix
    {
        private double[,] matrix;
        private double? cashedDeterminant;

        public MyMatrix(MyMatrix other)
        {
            matrix = new double[other.Height, other.Width];
            for (int i = 0; i < Height; i++)
                for (int j = 0; j < Width; j++)
                    matrix[i, j] = other[i, j];
        }

        public MyMatrix(double[,] array)
        {
            matrix = new double[array.GetLength(0), array.GetLength(1)];
            for (int i = 0; i < Height; i++)
                for (int j = 0; j < Width; j++)
                    matrix[i, j] = array[i, j];
        }

        public MyMatrix(double[][] jaggedArray)
        {
            ValidateJaggedArray(jaggedArray);
            int width = jaggedArray[0].Length;
            matrix = new double[jaggedArray.Length, width];
            for (int i = 0; i < Height; i++)
                for (int j = 0; j < Width; j++)
                    matrix[i, j] = jaggedArray[i][j];
        }

        public MyMatrix(string[] rows)
        {
            if (rows == null || rows.Length == 0)
                throw new ArgumentException("Array cannot be empty");

            var firstRow = rows[0].Split(new[] { ' ', '\t' }, StringSplitOptions.RemoveEmptyEntries);
            matrix = new double[rows.Length, firstRow.Length];

            for (int i = 0; i < rows.Length; i++)
            {
                var numbers = rows[i].Split(new[] { ' ', '\t' }, StringSplitOptions.RemoveEmptyEntries);
                if (numbers.Length != firstRow.Length)
                    throw new ArgumentException("All rows must have the same number of elements");

                for (int j = 0; j < numbers.Length; j++)
                    if (!double.TryParse(numbers[j], out matrix[i, j]))
                        throw new ArgumentException($"Invalid number format at row {i}, column {j}");
            }
        }

        public MyMatrix(string matrixString)
        {
            var rows = matrixString.Split(new[] { '\n', '\r' }, StringSplitOptions.RemoveEmptyEntries);
            if (rows.Length == 0)
                throw new ArgumentException("String cannot be empty");

            var firstRow = rows[0].Split(new[] { ' ', '\t' }, StringSplitOptions.RemoveEmptyEntries);
            matrix = new double[rows.Length, firstRow.Length];

            for (int i = 0; i < rows.Length; i++)
            {
                var numbers = rows[i].Split(new[] { ' ', '\t' }, StringSplitOptions.RemoveEmptyEntries);
                if (numbers.Length != firstRow.Length)
                    throw new ArgumentException("All rows must have the same number of elements");

                for (int j = 0; j < numbers.Length; j++)
                    if (!double.TryParse(numbers[j], out matrix[i, j]))
                        throw new ArgumentException($"Invalid number format at row {i}, column {j}");
            }
        }

        public int Height => matrix.GetLength(0);
        public int Width => matrix.GetLength(1);

        public int getHeight() => Height;
        public int getWidth() => Width;

        public double this[int i, int j]
        {
            get
            {
                ValidateIndexes(i, j);
                return matrix[i, j];
            }
            set
            {
                ValidateIndexes(i, j);
                cashedDeterminant = null;
                matrix[i, j] = value;
            }
        }

        public double getElement(int i, int j)
        {
            ValidateIndexes(i, j);
            return matrix[i, j];
        }

        public void setElement(int i, int j, double value)
        {
            ValidateIndexes(i, j);
            cashedDeterminant = null;
            matrix[i, j] = value;
        }

        private void ValidateIndexes(int i, int j)
        {
            if (i < 0 || i >= Height)
                throw new IndexOutOfRangeException($"Row index {i} is out of range [0, {Height})");
            if (j < 0 || j >= Width)
                throw new IndexOutOfRangeException($"Column index {j} is out of range [0, {Width})");
        }

        private void ValidateJaggedArray(double[][] jaggedArray)
        {
            if (jaggedArray == null || jaggedArray.Length == 0)
                throw new ArgumentException("Matrix cannot be empty");

            int width = jaggedArray[0].Length;

            for (int i = 1; i < jaggedArray.Length; i++)
                if (jaggedArray[i].Length != width)
                    throw new ArgumentException("Matrix must be rectangular");
        }

        public override string ToString()
        {
            var result = new System.Text.StringBuilder();
            for (int i = 0; i < Height; i++)
            {
                for (int j = 0; j < Width; j++)
                {
                    result.Append(matrix[i, j]);
                    if (j < Width - 1)
                        result.Append('\t');
                }
                if (i < Height - 1)
                    result.AppendLine();
            }
            return result.ToString();
        }
    }
}
